Improve color selection and display

Nerixyz opened the pull request:
<!--
    Please include a summary of what you've changed and what issue is fixed.
    In the case of a bug fix, please include steps to reproduce the bug, so the pull request can be tested.
    If this PR fixes an issue on GitHub, mention this here to automatically close it: "Fixes #1234.".
-->

This PR improves color selection and display by doing the following changes:

- A checkerboard is displayed where the color isn't opaque
- Switched to saturation-brightness canvas
- Added an alpha slider
- The CSS input now shows alpha and accepts `#rgb`, `#rrggbb`, and `#rrggbbaa` (this is the CSS one, not the Qt one)

| Before | After |
|---| ---|
| ![firefox_2023-12-30_16-29-55](https://github.com/Chatterino/chatterino2/assets/19953266/6012fae7-e61b-417c-9990-5d8cc24aa7b5) | ![chatterino_2023-12-30_16-27-01](https://github.com/Chatterino/chatterino2/assets/19953266/3120b298-9f86-4d51-ba17-aa1c46685436) |
| ![firefox_2023-12-30_16-31-08](https://github.com/Chatterino/chatterino2/assets/19953266/2cd98a9f-2e5c-448c-9b59-dbc01a10565c) | ![firefox_2023-12-30_16-28-57](https://github.com/Chatterino/chatterino2/assets/19953266/20687abf-ab66-49c8-b2b1-65dee40053e5) <br> same for users and badges |

I'm still not sure about the rounding of the widgets and the circle selector, so feedback is welcome!

pajlada commented:
Some quick feedback
1) At the current size, It looks like we should have room for at least one more row of recently used colors. Maybe we should add more recently used colors, or we should reduce the height of the dialog overall
1) I don't mind the roundedness of the recently used/default colors/selected color blobs
1) I don't mind the roundedness of the hue or alpha sliders
1) I like the new color widget thing in the highlight dialog
1) I miss the "Selected" label for the current color. I think it would be nice to have a similar box under it/next to it with the old color
1) The circle cursor would be good if it had a solid color of the actually selected color (e.g. how it's done here: https://coloris.js.org/examples.html). I understand that might be a bit more challenging to implement, so I'm fine with using a simple `+` cursor too

Nerixyz commented:
> At the current size, It looks like we should have room for at least one more row of recently used colors. Maybe we should add more recently used colors, or we should reduce the height of the dialog overall

The recent colors are just the colors of the message and user highlights. So if you have more entries with different colors, you'll have more recently used colors.

>     5. I miss the "Selected" label for the current color. I think it would be nice to have a similar box under it/next to it with the old color

Added.

> 6. The circle cursor would be good if it had a solid color of the actually selected color (e.g. how it's done here: [coloris.js.org/examples.html](https://coloris.js.org/examples.html)). I understand that might be a bit more challenging to implement, so I'm fine with using a simple `+` cursor too

I added the opaque color, but it doesn't really make much of a difference. At least in the alpha slider, I think it's better to be transparent. Doing it the way it's done on the web isn't as easy, as we need to add another mask when painting the checkerboard (which seems overkill).

pajlada commented:
Can revert the circle opaque color in that case, it was probably fine

pajlada commented on src/widgets/helper/color/ColorSelect.hpp:
```diff
@@ -0,0 +1,46 @@
+class ColorSelect : public QWidget
+{
+    Q_OBJECT
+    Q_PROPERTY(QColor color READ color WRITE setColor NOTIFY colorChanged)
```
Remove `Q_PROPERTY` usage here & in other files

pajlada commented on src/widgets/helper/color/ColorSelect.hpp:
```diff
@@ -0,0 +1,46 @@
+    QVBoxLayout rows_;
+
+    void connectWidgets();
+
+signals:
+    void pickerColorChanged(QColor color);
+    void detailsColorChanged(QColor color);
```
Remove these unused signals

pajlada commented on src/widgets/helper/color/SBCanvas.hpp:
```diff
@@ -0,0 +1,55 @@
+
+namespace chatterino {
+
+class SBCanvas : public QWidget
```
Comment on the class what it is since "SB" isn't a known acronym to me, or just rename it to its full name

pajlada commented on src/widgets/helper/color/SBCanvas.hpp:
```diff
@@ -0,0 +1,55 @@
+    void saturationChanged(int saturation) const;
+    void brightnessChanged(int brightness) const;
+
+public slots:
+    int saturation() const;
+    int brightness() const;
```
These should not be slots

pajlada commented on src/widgets/helper/color/ColorDetails.cpp:
```diff
@@ -0,0 +1,160 @@
+    this->updateCss();
+}
+
+void ColorDetails::updateCss()
```
Call it Hex instead as discussed in chat

pajlada commented on src/widgets/helper/color/ColorDetails.hpp:
```diff
@@ -0,0 +1,51 @@
+
+namespace chatterino {
+
+class ColorDetails : public QWidget
```
Class naming is iffy to me, maybe ColorInputs?

pajlada commented on src/widgets/helper/color/ColorSelect.cpp:
```diff
@@ -0,0 +1,86 @@
+
+void ColorSelect::connectWidgets()
+{
+    const auto applyColor = [this]() {
+        emit colorChanged(this->currentColor_);
+    };
```
Can we get rid of this lambda since it's a single-line anyway?

pajlada reviewed (commented):
If it's not the biggest hassle, I'd like to get rid of the `ColorSelect` "widget" (and maybe even the ColorDetails) and just add them straight to the ColorPickerDialog - going through the anatomy of the dialog has been a few too many classes deep imo, maybe that will make it a bit simpler

pajlada commented on src/widgets/helper/color/ColorButton.cpp:
```diff
@@ -0,0 +1,76 @@
+
+void ColorButton::setColor(const QColor &color)
+{
+    this->currentColor_ = color;
+    this->update();
```
Should we check to only call update if currentColor_ and color are not the same? or is that overengineering this use case

Nerixyz replied:
Yes we should check this.

pajlada reviewed (approved):
Looks good to me! Just one quick review question & I'm happy to merge this in!
